contracts/GNTPaymentChannels.sol

```diff
@@ -39,157 +33,114 @@ contract GNTPaymentChannels is ReceivingContract {
-        bytes32 channel = keccak256(id++);
-        channels[channel] = PaymentChannel(msg.sender, _receiver, 0, 0, 0);
-        emit NewChannel(msg.sender, _receiver, channel); // event
-    }
-
     // Fund existing channel; can be done multiple times.
-    function onTokenReceived(address _from, uint _value, bytes _data) public onlyToken {
-        bytes32 channel;
+    function onTokenReceived(address _owner, uint256 _value, bytes _data) public onlyToken {
+        require(_data.length == 20);
+        address receiver;
         assembly {
-          channel := mload(add(_data, 32))
+          receiver := mload(add(_data, 20))
         }
-        PaymentChannel storage ch = channels[channel];
-        require(_from == ch.owner);
+        PaymentChannel storage ch = _getChannel(_owner, receiver);
         ch.deposited += _value;
-        emit Fund(ch.receiver, channel, _value);
+        emit Fund(_owner, receiver, _value);
     }
 
     // Receiver can withdraw multiple times without closing the channel
     function withdraw(
-        bytes32 _channel,
+        address owner,
         uint256 _value,
         uint8 _v,
         bytes32 _r,
         bytes32 _s
     )
         external
-        onlyValidSig(_channel, _value, _v, _r, _s)
-        returns (bool)
+        onlyValidSig(owner, msg.sender, _value, _v, _r, _s)
     {
-        PaymentChannel storage ch = channels[_channel];
-        require(ch.withdrawn < _value); // <- STRICT less than!
+        address receiver = msg.sender;
+        PaymentChannel storage ch = _getChannel(owner, receiver);
+        require(ch.withdrawn < _value);
         uint256 amount = _value - ch.withdrawn;
         // Receiver has been cheated! Withdraw as much as possible.
-        if (ch.deposited - ch.withdrawn < amount)
+        if (ch.deposited - ch.withdrawn < amount) {
             amount = ch.deposited - ch.withdrawn;
-        return _do_withdraw(_channel, amount);
+        }
+
+        ch.withdrawn += amount;
+        require(token.transfer(receiver, amount));
+        emit Withdraw(owner, receiver);
     }
 
     // If receiver does not want to close channel, owner can do that
     // by calling unlock and waiting for grace period (close_delay).
-    function unlock(bytes32 _channel) external onlyOwner(_channel) {
-        PaymentChannel storage ch = channels[_channel];
+    function unlock(address receiver) external {
+        address owner = msg.sender;
+        PaymentChannel storage ch = _getChannel(owner, receiver);
         ch.locked_until = block.timestamp + close_delay;
-        emit TimeLocked(ch.owner, ch.receiver, _channel);
+        emit TimeLocked(owner, receiver);
     }
 
     // Owner can close channel to reclaim its money.
-    function close(bytes32 _channel)
-        external
-        onlyOwner(_channel)
-        unlocked(_channel)
-        returns (bool)
-    {
-        return _do_close(_channel, false);
-    }
+    function close(address receiver) external {
```

Comment:
Please add a clear comments that this function does not delete fields and there is no use for closing a channel if `withdraw == deposited` - it is actually a withdrawal by the owner.